lib/ecto/repo/queryable.ex

```diff
@@ -485,4 +541,31 @@ defmodule Ecto.Repo.Queryable do
       query: query,
       message: "expected a from expression with a schema"
   end
+
+  defp assert_structs!([head | _] = structs) when is_list(structs) do
+    with  [] <- Enum.reject(structs, &schema?/1),
+         true <- Enum.all?(structs, &(&1.__struct__ == head.__struct__)) do
+      :ok
+    else
+      rejected when is_list(rejected) ->
+        raise ArgumentError, "expected a struct or a list of structs, received #{inspect(rejected)}"
+
+      false ->
+        raise ArgumentError, "expected an homogenous list, received different struct types"
+    end
```

Comment:
This can be clearer as conditionals:

```elixir
if not Enum.all?(structs, &schema?/1) do
  raise ArgumentError, "expected a struct or a list of structs, received #{inspect(rejected)}"
end

if not Enum.all?(structs, &(&1.__struct__ == head.__struct__)) do
  raise ArgumentError, "expected an homogenous list, received different struct types"
end

:ok
end
```